### PR TITLE
[tsan] Add instrumentation to caml_call_gc

### PR DESCRIPTION
**Pull request by ghost**
This is the fix for issue #13040, as discussed there, and ported to the remaining backends.

Tested on all platforms with a native compiler, no regression observed and the `statmemprof/exception_comballoc.ml` test now passes.

### CODE REVIEW

**Comment by OlivierNicole** [comment]
(Just enabling the TSan CI for extra certainty, for some reason it’s better at observing bugs that are silent on my machine.)

**Comment by ghost** [comment]
> (Just enabling the TSan CI for extra certainty, for some reason it’s better at observing bugs that are silent on my machine.)

`s/bug/lamb/` then.

**Review by gasche — APPROVED** [review]
I believe that this is correct.

**Comment by ghost** [comment]
Of course this will need to be backported to 5.2.

**Review by OlivierNicole — APPROVED** [review]
LGTM too.

**Review by fabbing — APPROVED** [review]

File: Changes [hunk 1]
```diff
@@ -148,6 +148,12 @@ _______________
   The syntax '-(1 [@foo])' was incorrectly parsed as '-1'.
   (Jules Aguillon, reviewed by Gabriel Scherer, report by Gabriel Scherer)
 
+- #13058: Add TSan instrumentation to caml_call_gc(), since it may raise
+  exceptions.
+  (Fabrice Buoro, Olivier Nicole, Gabriel Scherer and Miod Vallat, review
+  by TBD).
```
Comment (gasche) [inline Changes]: I guess in this case you can remove the "review by ..." field, given that this is more of a collaborative effort with self-reviews.

Comment (gasche) [reply Changes]: If you want this in 5.2, you should move it in the 5.2.0 section right away.

Comment (ghost) [reply Changes]: Done.

**Comment by gasche** [comment]
@dra27 we got a flexlink-related failure on AppVeyor, that seems unrelated to the present PR:

```
Building flexlink.exe with TOOLCHAIN=mingw for OCaml 5.3.0
rm -f flexlink.exe
../boot/ocamlrun.exe ../boot/ocamlc -use-prims ../runtime/primitives -nostdlib -I ../stdlib -o flexlink.exe -cclib "version_res.o" version.ml Compat.ml coff.ml cmdline.ml create_dll.ml reloc.ml
x86_64-w64-mingw32-gcc -c -DMINGW -o flexdll_mingw64.o flexdll.c
x86_64-w64-mingw32-gcc -c -o flexdll_initer_mingw64.o flexdll_initer.c
make[3]: Leaving directory '/cygdrive/c/projects/🐫реализация-mingw64/flexdll'
cp flexdll/flexlink.exe flexlink.byte.exe
mkdir -p byte/bin
rm -f byte/bin/flexlink.exe
cp boot/ocamlrun.exe byte/bin/flexlink.exe
cat flexlink.byte.exe >> byte/bin/flexlink.exe
cp flexdll/flexdll_mingw64.o flexdll/flexdll_initer_mingw64.o byte/bin
  MKEXE runtime/ocamlrun.exe
/usr/lib/gcc/x86_64-w64-mingw32/11/../../../../x86_64-w64-mingw32/bin/ld: cannot find "C:\projects\🐫реализация-mingw64\byte\bin\flexdll_mingw64.o": No such file or directory
collect2: error: ld returned 1 exit status
** Fatal error: Error during linking
```

**Comment by gasche** [comment]
@OlivierNicole if you would like to take care to merge this one, it is an occasion to learn about cherry-picking in the release branch. (There is documentation in HACKING.adoc on our recommend flags for cherry-picking.) Let me know if you don't have time, and I would take care of it.

**Comment by OlivierNicole** [comment]
Hope I did not mess that up.

**Comment by gasche** [comment]
Looks fine to me, thanks!

**Pull request merged**